### PR TITLE
Adding DoctrineClearEntityManagerWorkerSubscriber to reset EM in worker

### LINKED ISSUE #34073: [Messenger] Use Doctrine middleware by default
**Description**  
After #34066, the Doctrine middleware should now work nicely.

In fact, I think that the "clear entity manager" and "ping connection" middleware should ALWAYS be used (assuming Doctrine is installed). Both of these are very sensible things to do to make sure the database connection is alive and that data in the entity manager isn't "stale" because it's being stored in an identity map that was added to via a message handled 10 minutes ago.

There are 2 tricky parts to this:

1) The services for those middleware are added via DoctrineBundle... but the default middleware need to be added in FrameworkExtension.

2) Would we do it for *all* entity managers? I think that could (in https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bridge/Doctrine/Messenger/DoctrinePingConnectionMiddleware.php) cause connections to be made to *all* connections, even if a connection wasn't previously open (though, that could be fixed by only pinging a connection `if $connection->isConnected()`).

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes & no :) 
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #34073
| License       | MIT
| Doc PR        | symfony/symfony-docs#12575
| DoctrineBundle PR | doctrine/DoctrineBundle#1043

Hi!

I've seen a few developers get "bit" by an issue recently: after running `messenger:consume`, the 2nd, 3rd, 4th, etc message that are handled are getting out-of-date data from Doctrine. The reason is simple:

A) Consume the 1st message, it queries for `Foo id=1` to do something
B) 10 minutes go by
C) Consume the 2nd message. It also queries for `Foo id=1`, but because this is already in the identity map, Doctrine re-uses the data that is now *10* minutes old. 

Even though one worker process handles many messages, the system should (as much as possible) isolate each handler from each other. This is one very practical place we can help people. Also, checking the code, I don't think clearing the entity manager will cause any issues for an EM whose Connection has not been "connected" yet (i.e. it will not cause a connection to be established).

We would wire this in DoctrineBundle, and could make it disable-able... in case that's something that's needed.

### CODE REVIEW

File: src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    public static function getSubscribedEvents()
+    {
+        yield WorkerMessageHandledEvent::class => 'onWorkerMessageHandled';
+        yield WorkerMessageFailedEvent::class => 'onWorkerMessageFailed';
+    }
+
+    private function clearEntityManagers()
+    {
+        foreach ($this->managerRegistry->getManagerNames() as $managerName) {
```
Comment (bendavies) [inline src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php]:
```php
        foreach ($this->managerRegistry->getManagers() as $manager) {
            $manager->clear();
        }
```

File: src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        $this->clearEntityManagers();
+    }
+
+    public function onWorkerMessageFailed()
+    {
+        $this->clearEntityManagers();
+    }
+
+    public static function getSubscribedEvents()
```
Comment (bendavies) [inline src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php]: could be simplified to clear before handling with just `WorkerMessageReceivedEvent`

**Comment by Tobion** [comment]
Seems fine to me to add this.

File: src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php (hunk 2 above)
Comment (weaverryan) [reply src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php]: You're totally right. What I don't know is: is doing it before vs after significant? I was following the "lead" of the middleware... but I'm not sure if it matters or not.

**Comment by bendavies** [comment]
👍 I have pretty much exactly this in a project

File: src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php (hunk 2 above)
Comment (bendavies) [reply src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php]: "after" will probably free up memory sooner, in the case they no messages come in for a while.

Comment (weaverryan) [reply src/Symfony/Bridge/Doctrine/Messenger/DoctrineClearEntityManagerWorkerSubscriber.php]: The fact that I need to listen to 2 events relates to #32614, which attempts to fix that.

**Comment by Koc** [comment]
@weaverryan should doctrine middleware for same functionality be removed? Because now we are having 2 approaches to do same thing

**Comment by weaverryan** [comment]
> @weaverryan should doctrine middleware for same functionality be removed? Because now we are having 2 approaches to do same thing

Excellent question :). You probably know better than me. I personally think this should be done *always* and has no downside (well, a few micro-seconds added after handling each message to clear)... so there would indeed be no use-case anymore for the "clear" middleware. Do you agree?

**Comment by Koc** [comment]
yes, event subscriber completely replaces middleware, so we can remove it.

**Comment by weaverryan** [comment]
DoctrineClearEntityManagerMiddleware is gone and docs PR added

**Comment by weaverryan** [comment]
Failures unrelated - coming from 4.4 branch

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Pull request merged**